validations/checkVariableAssignedBeforeUsed.js

```diff
@@ -0,0 +1,130 @@
+ * @param {Set} assignedVars - Set of variables that have been assigned so far
+ */
+function checkVariableInComputationHelper(
+  engine,
+  template,
+  assignedVars,
+  errorsArr
+) {
+  let currentSet = new Set(assignedVars);
+  let curErrors = [];
+
+  if (template.name === "if") {
+    handleIf(engine, template, currentSet, errorsArr);
+  } else if (template.name === "for" || template.name === "unless") {
+    handleForOrUnless(engine, template, currentSet, errorsArr);
+  } else if (template.name === "assign") {
+    const modifiedAssignedVars = handleAssign(
+      engine,
+      template,
+      curErrors,
+      assignedVars,
+      currentSet
+    );
+    modifiedAssignedVars.forEach((item) => currentSet.add(item));
+  }
+
+  errorsArr.push(...curErrors);
+}
+
+/**
+ * - Get variable used in if condition and add it to predefined vars set
+ * - Ex: for the case if (x===some conditon), adds x to set
+ * - passes this set to the next conditional branches
+ * @param {*} currentSet - set of all variables defined up until the current if block
+ */
+function handleIf(engine, template, currentSet, errorsArr) {
+  const conditionVar = template.token?.args;
+  if (conditionVar) currentSet.add(conditionVar);
+
+  (template.tagImpl.branches ?? []).forEach((branch) => {
+    currentSet.add(branch.cond[0]);
+    branch.templates.forEach((tpl) => {
+      if (tpl.type === "tag") {
+        isTemplateTypeTag(engine, tpl, currentSet, errorsArr);
+      }
+    });
+  });
+
+  (template.tagImpl.elseTemplates ?? []).forEach((tpl) => {
+    isTemplateTypeTag(engine, tpl, currentSet, errorsArr);
+  });
+}
+
+/**
+ * - Get variable used in if condition and add it to predefined vars set
+ * - Ex: for the case if (x===some conditon), adds x to set
+ * - passes this set to the next conditional branches
+ * @param {*} currentSet - set of all variables defined up until the current if block
+ */
+function handleForOrUnless(engine, template, currentSet, errorsArr) {
+  (template.tagImpl?.templates ?? []).forEach((tpl) => {
+    isTemplateTypeTag(engine, tpl, currentSet, errorsArr);
+  });
+
+  (template.tagImpl?.elseTemplates ?? []).forEach((tpl) => {
+    isTemplateTypeTag(engine, tpl, currentSet, errorsArr);
+  });
+}
+
+/**
+ * @param {*} assignedVars - set of all variables defined before the assign statement
+ */
+function handleAssign(engine, template, curErrors, assignedVars) {
+  const parsedObj = depGraph.parseAssign(template, engine);
+
+  (parsedObj.dependsOn ?? []).forEach((varName) => {
+    if (!assignedVars.has(varName)) {
+      curErrors.push(
+        `Variable "${varName}" used before assignment in expression "${template.token.args}" on line ${template.token.line}`
+      );
+    }
+  });
+
+  if (curErrors.length === 0) {
+    assignedVars.add(parsedObj.defined);
+  }
+  return assignedVars;
+}
+
+/**
+ * - recursively passes tag to helper function to process it further
+ */
+function isTemplateTypeTag(engine, tpl, currentSet, errorsArr) {
```

Comment:
The name of the function is `isTemplateTypeTag`,  which means that it should return a boolean whether the passed template.type is tag or not. We're doing some other stuff also inside.

Let this method just return a boolean. This can be used in multiple places in this file.

```js
function isTemplateTypeTag(tpl) {
  return tpl.type === "tag";
}
```

You can create another function for the other stuff?
```js
function handleTag(engine, tpl, currentSet, errorsArr) {
  if (isTemplateTypeTag(tpl)) {
    checkVariableInComputationHelper(engine, tpl, currentSet, errorsArr);
  }
}
```